### ISSUE TITLE
npm installation hangs up / yarn  Segmentation fault

### ISSUE THREAD

**Issue by dadeppa** [jishi/node-sonos-http-api#518]
Hi,

I tested this api with the prebuild image and it worked fine on an raspberry. Now I wanted to transfer it to my server / NAS, but unfortunately it does not work.

I am running an XPenology on an HP MicroServer, system is running pretty well with DSM 5.2. I managed to install ipkg and node.

My versions are:

```
ds> node -v
v6.7.0
ds> npm -v
3.10.3
ds> yarn version
yarn version v0.24.5

```

I first try to install with npm, but It froze at this point:
```
ds> npm install
[    ..............] / fetchMetadata: sill mapToRegistry uri https://registry.n

```
Then I tried to follow an attemp using yarn to install it, this responded to me with the following error:

```
ds> yarn install
yarn install v0.24.5
info No lockfile found.
[1/4] Resolving packages...
⠄ sonos-discovery@https://github.com/jishi/node-sonos-discovery/archive/v1.2.5. Segmentation fault

```

I also tried with the --production behind npm and tried to start the app directly - which was not working. 

I have no proxy for HTTP and HTTPS in my node config.

So here is where my skills end. I hope someone can help me or point to the right direction.

Thank you,

Andreas

**Comment by jishi** [comment]
How did you download the source? It tries to download sonos-discovery 1.2.5, but the latest version should try and download 1.4.1... 

I haven't tested installing with yarn, but it should work equally well. It sounds more like it can't download the tar.gz dependency of sonos-discovery for some reason. 

It could also be a temporary problem with the npm registry. It's not uncommon that they have operational issues form time to time.

**Comment by dadeppa** [comment]
Thanks for your reply.

As I am no expert on linux I will describe what I did:

After struggling around with installing node - which finally worked - I tried to install your api on different ways;

First I went to your releases ( https://github.com/jishi/node-sonos-http-api/releases ) and downloaded the zip file v 1.2.6, unzipped it, transferred it to my server and then I did some linux things; cd'd to the folder where I put it and ran 'npm install --production'.

As this did not work, I searched for the folder, where the modules are stored (do not know which it is actually), cd'd there and did 'npm install https://github.com/jishi/node-sonos-http-api/archive/v1.2.6.tar.gz' - which did not work either.

Then I installed yarn and did the same with yarn as with npm - with the errors posted above.

Shall I install all dependencies separately?

**Comment by jishi** [comment]
Hm, okay. I only do releases when I deem it necessary, since that would auto-update all installations based on the raspberry pi automatically.

Try this: Download the latest master https://github.com/jishi/node-sonos-http-api/archive/master.zip unzip it and then run `npm install --production` to see if that works.

Also make sure that you can actually download https://github.com/jishi/node-sonos-discovery/archive/v1.4.1.tar.gz on your NAS, there might be something broken with your https request (missing certificates on your machines) or something, that might be the cause of it.

**Comment by dadeppa** [comment]
Ok, I think I understand now, what is a difference between release and the code section ...

I followed your instructions, unfortunately, it didn't work (still hanging up on `npm install --production` after cd'ing to the directory.

I then downloaded node-sonos-discovery, moved it to the NAS and tried to install...

When I run `npm install --production` it will freeze, but when I install it as global `npm install -g --production` it says:
`> npm install --production -g
/volume1/@entware/lib
└── sonos-discovery@1.4.1
`
where the sonos-discovery is yellow... is this a good sign?

I tried to check the installation:
`ds> npm ls
sonos-discovery@1.4.1 /volume1/Software/node-sonos-discovery-master
├── UNMET DEPENDENCY html-entities@1.0.x
└── UNMET DEPENDENCY xml-flow@^1.0.1

npm ERR! missing: html-entities@1.0.x, required by sonos-discovery@1.4.1
npm ERR! missing: xml-flow@^1.0.1, required by sonos-discovery@1.4.1
`

So i guess still dependences missing... 

I tried to install them by name as global:
`ds> npm install html-entities -g --production
/volume1/@entware/lib
└── html-entities@1.2.1

ds> npm install xml-flow -g --production
/volume1/@entware/lib
└─┬ xml-flow@1.0.1
  └── sax@1.2.2

ds> npm ls
sonos-discovery@1.4.1 /volume1/Software/node-sonos-discovery-master
├── UNMET DEPENDENCY html-entities@1.0.x
└── UNMET DEPENDENCY xml-flow@^1.0.1

npm ERR! missing: html-entities@1.0.x, required by sonos-discovery@1.4.1
npm ERR! missing: xml-flow@^1.0.1, required by sonos-discovery@1.4.1`

... so what am I still doing wrong?

**Comment by jishi** [comment]
does a regular `npm install xml-flow` in the `node-sonos-http-api` work for you? Like, manually installing dependencies without the global flag? 

It's not really recommended to install dependencies globally, it might give you a headache in the future (especially when upgrading). It could be some permissions wonkyness, but to be honest, this is a very weird edge case that I've never heard of. 

Even if you download sonos-discovery manually, the dependencies for that package needs to be fulfilled so it will be a pain to get it going manually if the npm install doesn't work. What you can do, is unzip and install the dependencies (run `npm install`) on another machine, and then copy everything (including the `node_modules` folder) to your NAS. I don't think I have any modules that requires platform specific compiling, so it should work on any other machine that has node/npm installed.

But investigate the permission theory to start with. As a baseline test, give rw permissions to everything in the node-sonos-htt-api folder that you create on your NAS, recursively.

**Comment by dadeppa** [comment]
Okay, I first tried to check the rights-guess:

I downloaded the Code as zip, extracted it to my NAS to /volume1/Software/node-sonos-http-api/

Then I set chmod 755 recursive and did the following:
`ds> cd /volume1/Software/node-sonos-http-api
ds> ls
@eaDir             README.md          presets            static
ISSUE_TEMPLATE.md  lib                server.js          test_endpoint.js
LICENSE.md         package.json       settings.js
ds> ls -al
drwxr-xr-x    6 admin    users         4096 Jun 15 00:06 .
drwxrwxrwx   39 root     root          4096 Jun 15 00:07 ..
-rwxr-xr-x    1 admin    users           80 Jun 12 12:13 .gitattributes
-rwxr-xr-x    1 admin    users          231 Jun 12 12:13 .gitignore
drwxr-xr-x    2 root     users         4096 Jun 15 00:06 @eaDir
-rwxr-xr-x    1 admin    users          750 Jun 12 12:13 ISSUE_TEMPLATE.md
-rwxr-xr-x    1 admin    users         1079 Jun 12 12:13 LICENSE.md
-rwxr-xr-x    1 admin    users        28954 Jun 12 12:13 README.md
drwxr-xr-x    7 admin    users         4096 Jun 15 00:06 lib
-rwxr-xr-x    1 admin    users          776 Jun 12 12:13 package.json
drwxr-xr-x    3 admin    users         4096 Jun 15 00:06 presets
-rwxr-xr-x    1 admin    users         2964 Jun 12 12:13 server.js
-rwxr-xr-x    1 admin    users         1252 Jun 12 12:13 settings.js
drwxr-xr-x    5 admin    users         4096 Jun 15 00:06 static
-rwxr-xr-x    1 admin    users          512 Jun 12 12:13 test_endpoint.js
ds> npm install --production
[    ..............] - fetchMetadata: sill mapToRegistry uri https://registry.npmjs.org/request-promise`


There it freezes again :'( 

The regular `npm install xml-flow` worked:

`datenhafen> npm install xml-flow
/volume1/Software/sonos
└─┬ xml-flow@1.0.1
  └── sax@1.2.2

npm WARN enoent ENOENT: no such file or directory, open '/volume1/Software/sonos/package.json'
npm WARN sonos No description
npm WARN sonos No repository field.
npm WARN sonos No README data
npm WARN sonos No license field.
`

... time to throw in the towel?

**Comment by jishi** [comment]
Hm, but why are you invoking `npm install` in folder `/volume1/Software/sonos` and not `/volume1/Software/node-sonos-http-api`?

I mean, of course you can install every individual dependency manually like that (just look in the package.json to get a list of required packages). Still unsure why npm install fails. I know it can be slow though, you should at least give it 2-3 minutes to complete (but I assume that something else is wrong because yarn fails completely).

**Comment by dadeppa** [comment]
Yeah, i switched to a new folder (sonos), to get everything started at zero.

I in advance did a compete new install of an alpine on my raspberry pi 1, then installed node and npm and installed... tadaaaaaa... node-sonos-http-api... unfortunately it deletes itself after a reboot. So I downloaded your premade image, killed the running server, created a new folder and installed it there ... working! 

So I guess this is an itsy-bitsy-shitty configuration thing on my XPenology System.

The things I tried then back on the XPenology system:

- Clear folder 'sonos'
- Try to install:
`ds> pwd
/volume1/Software/sonos
ds> npm install jishi/node-sonos-http-api --production
npm WARN addRemoteGit Error: not found: git
npm WARN addRemoteGit     at getNotFoundError (/volume1/@entware/lib/node_module                                                                                                                                                             s/npm/node_modules/which/which.js:14:12)
npm WARN addRemoteGit     at F (/volume1/@entware/lib/node_modules/npm/node_modu                                                                                                                                                             les/which/which.js:69:19)
npm WARN addRemoteGit     at E (/volume1/@entware/lib/node_modules/npm/node_modu                                                                                                                                                             les/which/which.js:81:29)
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/which.js:90:16
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/node_modules/isexe/index.js:44:5
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/node_modules/isexe/access.js:8:5
npm WARN addRemoteGit     at FSReqWrap.oncomplete (fs.js:123:15)
npm WARN addRemoteGit  jishi/node-sonos-http-api resetting remote /root/.npm/_gi                                                                                                                                                             t-remotes/git-github-com-jishi-node-sonos-http-api-git-9f098f57 because of error                                                                                                                                                             : { Error: not found: git
npm WARN addRemoteGit     at getNotFoundError (/volume1/@entware/lib/node_module                                                                                                                                                             s/npm/node_modules/which/which.js:14:12)
npm WARN addRemoteGit     at F (/volume1/@entware/lib/node_modules/npm/node_modu                                                                                                                                                             les/which/which.js:69:19)
npm WARN addRemoteGit     at E (/volume1/@entware/lib/node_modules/npm/node_modu                                                                                                                                                             les/which/which.js:81:29)
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/which.js:90:16
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/node_modules/isexe/index.js:44:5
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/node_modules/isexe/access.js:8:5
npm WARN addRemoteGit     at FSReqWrap.oncomplete (fs.js:123:15) code: 'ENOGIT'                                                                                                                                                              }
npm ERR! git clone --template=/root/.npm/_git-remotes/_templates --mirror git://                                                                                                                                                             github.com/jishi/node-sonos-http-api.git /root/.npm/_git-remotes/git-github-com-                                                                                                                                                             jishi-node-sonos-http-api-git-9f098f57: undefined
npm ERR! git clone --template=/root/.npm/_git-remotes/_templates --mirror git://                                                                                                                                                             github.com/jishi/node-sonos-http-api.git /root/.npm/_git-remotes/git-github-com-                                                                                                                                                             jishi-node-sonos-http-api-git-9f098f57: undefined
npm WARN addRemoteGit Error: not found: git
npm WARN addRemoteGit     at getNotFoundError (/volume1/@entware/lib/node_module                                                                                                                                                             s/npm/node_modules/which/which.js:14:12)
npm WARN addRemoteGit     at F (/volume1/@entware/lib/node_modules/npm/node_modu                                                                                                                                                             les/which/which.js:69:19)
npm WARN addRemoteGit     at E (/volume1/@entware/lib/node_modules/npm/node_modu                                                                                                                                                             les/which/which.js:81:29)
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/which.js:90:16
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/node_modules/isexe/index.js:44:5
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/node_modules/isexe/access.js:8:5
npm WARN addRemoteGit     at FSReqWrap.oncomplete (fs.js:123:15)
npm WARN addRemoteGit  jishi/node-sonos-http-api resetting remote /root/.npm/_gi                                                                                                                                                             t-remotes/git-https-github-com-jishi-node-sonos-http-api-git-59bf2b22 because of                                                                                                                                                              error: { Error: not found: git
npm WARN addRemoteGit     at getNotFoundError (/volume1/@entware/lib/node_module                                                                                                                                                             s/npm/node_modules/which/which.js:14:12)
npm WARN addRemoteGit     at F (/volume1/@entware/lib/node_modules/npm/node_modu                                                                                                                                                             les/which/which.js:69:19)
npm WARN addRemoteGit     at E (/volume1/@entware/lib/node_modules/npm/node_modu                                                                                                                                                             les/which/which.js:81:29)
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/which.js:90:16
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/node_modules/isexe/index.js:44:5
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/node_modules/isexe/access.js:8:5
npm WARN addRemoteGit     at FSReqWrap.oncomplete (fs.js:123:15) code: 'ENOGIT'                                                                                                                                                              }
npm ERR! git clone --template=/root/.npm/_git-remotes/_templates --mirror https:                                                                                                                                                             //github.com/jishi/node-sonos-http-api.git /root/.npm/_git-remotes/git-https-git                                                                                                                                                             hub-com-jishi-node-sonos-http-api-git-59bf2b22: undefined
npm ERR! git clone --template=/root/.npm/_git-remotes/_templates --mirror https:                                                                                                                                                             //github.com/jishi/node-sonos-http-api.git /root/.npm/_git-remotes/git-https-git                                                                                                                                                             hub-com-jishi-node-sonos-http-api-git-59bf2b22: undefined
npm WARN addRemoteGit Error: not found: git
npm WARN addRemoteGit     at getNotFoundError (/volume1/@entware/lib/node_module                                                                                                                                                             s/npm/node_modules/which/which.js:14:12)
npm WARN addRemoteGit     at F (/volume1/@entware/lib/node_modules/npm/node_modu                                                                                                                                                             les/which/which.js:69:19)
npm WARN addRemoteGit     at E (/volume1/@entware/lib/node_modules/npm/node_modu                                                                                                                                                             les/which/which.js:81:29)
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/which.js:90:16
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/node_modules/isexe/index.js:44:5
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/node_modules/isexe/access.js:8:5
npm WARN addRemoteGit     at FSReqWrap.oncomplete (fs.js:123:15)
npm WARN addRemoteGit  jishi/node-sonos-http-api resetting remote /root/.npm/_gi                                                                                                                                                             t-remotes/git-github-com-jishi-node-sonos-http-api-git-10f8f145 because of error                                                                                                                                                             : { Error: not found: git
npm WARN addRemoteGit     at getNotFoundError (/volume1/@entware/lib/node_module                                                                                                                                                             s/npm/node_modules/which/which.js:14:12)
npm WARN addRemoteGit     at F (/volume1/@entware/lib/node_modules/npm/node_modu                                                                                                                                                             les/which/which.js:69:19)
npm WARN addRemoteGit     at E (/volume1/@entware/lib/node_modules/npm/node_modu                                                                                                                                                             les/which/which.js:81:29)
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/which.js:90:16
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/node_modules/isexe/index.js:44:5
npm WARN addRemoteGit     at /volume1/@entware/lib/node_modules/npm/node_modules                                                                                                                                                             /which/node_modules/isexe/access.js:8:5
npm WARN addRemoteGit     at FSReqWrap.oncomplete (fs.js:123:15) code: 'ENOGIT'                                                                                                                                                              }
npm ERR! git clone --template=/root/.npm/_git-remotes/_templates --mirror git@gi                                                                                                                                                             thub.com:jishi/node-sonos-http-api.git /root/.npm/_git-remotes/git-github-com-ji                                                                                                                                                             shi-node-sonos-http-api-git-10f8f145: undefined
npm ERR! git clone --template=/root/.npm/_git-remotes/_templates --mirror git@gi                                                                                                                                                             thub.com:jishi/node-sonos-http-api.git /root/.npm/_git-remotes/git-github-com-ji                                                                                                                                                             shi-node-sonos-http-api-git-10f8f145: undefined
npm ERR! Linux 3.10.35
npm ERR! argv "/volume1/@entware/bin/node" "/opt/bin/npm" "install" "jishi/node-                                                                                                                                                             sonos-http-api" "--production"
npm ERR! node v6.7.0
npm ERR! npm  v3.10.3
npm ERR! code ENOGIT

npm ERR! not found: git
npm ERR!
npm ERR! Failed using git.
npm ERR! This is most likely not a problem with npm itself.
npm ERR! Please check if you have git installed and in your PATH.

npm ERR! Please include the following file with any support request:
npm ERR!     /volume1/Software/sonos/npm-debug.log`

- Checked for git -> It is installed as global package
- Installed git locally
`ds> npm install git
/volume1/Software/sonos
└─┬ git@0.1.5
  └── mime@1.2.9
`

- Then tried again with same erros (could not find git)

So and here I am screwed. As mentoined yesterday I am not that into linux and shell, but I got it managed to install on your and on a new alpine installation - but not on my NAS. 

So it seems, that I have to either search help in the XPenology forum or run another hardware raspberry pi to run the node-sonos. Maybe I will then even do the HMI of my home automation on the pi with its node.js on it and disable node.js on the NAS.

**Comment by dadeppa** [comment]
Okay... found an error: I did not install git via ipkg... that's what the error is telling me.

I Installed git and got other errors, which I solved by setting:
`ds> git -c core.longpaths=true config --get remote.origin.url`

But now it freezes at this point:
```
ds> npm install jishi/node-sonos-http-api --production
[    ..............] / fetchMetadata: sill mapToRegistry uri https://registry.npmjs.org/request-promise

```

**Comment by jishi** [comment]
You shouldn't need git to install it from a zip. That is only if you want to download the source using git. 

The reason for you getting errors about git was because you used `npm install jishi/node-sonos-http-api --production` which makes npm look for a git repository for user `jishi` named `node-sonos-http-api`. That's not what you want to do.

You need to take the zip, unzip it, and then stand in the folder, where there is a `package.json`, (i.e the root of the unzipped content), then run `npm install --production`. 

It's unclear exactly what steps you have tried, and you might have already tried it, but that's the basic gist of it.

**Comment by dadeppa** [comment]
Sorry for answering that late...

I did now download the zip, extraced it to my NAS and there I tried to install with --production. Unfortunately this still / also hangs at this point: 
[    ..............] / fetchMetadata: sill mapToRegistry uri https://registry.npmjs.org/request-promise

So I am still thinking, that this is an problem of my os, in this case XPenology / DSM. As I mentioned earlier, I will run this on a separate device, in this case a rpi 3, because with an own setup alpine linux this is all running fine...

Thank you for putting that much effort in explaining to me.
